codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java

```diff
@@ -368,22 +368,28 @@ private boolean allPromisesDone() {
         }
 
         private ChannelPromise setPromise() {
-            if (lastFailure == null) {
+            if (aggregateFailure == null) {
                 promise.setSuccess();
                 return super.setSuccess(null);
             } else {
-                promise.setFailure(lastFailure);
-                return super.setFailure(lastFailure);
+                promise.setFailure(aggregateFailure);
+                return super.setFailure(aggregateFailure);
             }
         }
 
         private boolean tryPromise() {
-            if (lastFailure == null) {
+            if (aggregateFailure == null) {
                 promise.trySuccess();
                 return super.trySuccess(null);
             } else {
-                promise.tryFailure(lastFailure);
-                return super.tryFailure(lastFailure);
+                promise.tryFailure(aggregateFailure);
+                return super.tryFailure(aggregateFailure);
+            }
+        }
+
+        private void setAggregateFailure(Throwable cause) {
+            if (aggregateFailure == null) {
+                aggregateFailure = cause;
             }
```

Comment:
Should we try to add it as suppressed via ThrowableUtil if we already have one stored ?

Reply:
first exception should be sufficient but we can do this if you think it adds value. `addSuppressed` will cause issues with static exceptions (e.g memory leak) and self-references (throws synchronously) so care is required when using this API (e.g. consider enhancing `ThrowableUtils` to avoid adding self-references).

Reply:
@slandelle WDYT ?

Reply:
@Scottmitch trying to add a self-reference will throw as Throwable itself guard against it. So I think there is nothing to do in `ThrowableUtil` for us here... I am missing anything ?

Reply:
> Throwable itself guard against it.

That is what I mean. Throwable's "guard" is to throw in this case, but we generally don't know if the same reference is being used or not. If the same reference in being used and we allow a throw here the async control flow may not  complete (e.g. `Promise.setSuccess` throws before it notifies listeners). For our usages where we don't control the exceptions and need to complete async control flow we should prevent passing a self-reference to `addSuppressed`.